packages/editor/src/components/provider/use-block-editor-settings.js

```diff
@@ -290,6 +293,11 @@ function useBlockEditorSettings( settings, postType, postId, renderingMode ) {
 			isDistractionFree,
 			keepCaretInsideBlock,
 			mediaUpload: hasUploadPermissions ? mediaUpload : undefined,
+			__experimentalMediaSideload: hasUploadPermissions
+				? mediaSideload
+				: undefined,
+			__experimentalValidateFileSize: validateFileSize,
+			__experimentalValidateMimeType: validateMimeType,
```

Comment:
Why are these two settings needed? Are they WP specific?

Reply:
No, they are not WordPress specific.

These separate functions are needed so that file size and mime type can be validated before a file is added to the queue.

Currently, this validation is only done in the `uploadMedia()` function from `media-utils`, which happens right before the file is sent to the server. But now there is additional processing before the uploading, so the two steps need to be separated.

Example 1 (file is too big):

1. Drop a 10GB image into the editor
2. File size is validated
3. File is too big and is rejected, nothing is added to the queue

Example 2 (file is OK):

1. Drop image into the editor
2. File size is validated
3. Mime type is validated
4. File is added to the queue, where its size and mime type may change (compression, conversion)
5. File is uploaded by calling `uploadMedia()`, which does once again validate file size and mime type under the hood

Since the upload queue may change a file's type or size in the future (e.g. for converting HEIF images), eventually the validation will look more like this:

https://github.com/swissspidy/media-experiments/blob/7b91599369f1d3ae412e3e18a4afe2f62257b4dc/packages/editor/src/init/editor/media-upload.ts#L61-L87

So the mime type would technically be disallowed, but since we know we can convert the image to an allowed mime type, it can be added to the queue nonetheless.

Reply:
Thanks for the details, it makes sense. Now, I do wonder why we need the settings though.

AFAIK the `block-editor` package already receives `allowedMimeTypes` and `maxUploadFileSize` settings. Why can't we bundle these within the packages directly instead of having these provided as extra settings?

Reply:
I think we should use private symbols for new experimental settings

Reply:
> AFAIK the `block-editor` package already receives `allowedMimeTypes` and `maxUploadFileSize` settings. Why can't we bundle these within the packages directly instead of having these provided as extra settings?

This way we avoid repetition. Otherwise we'd have to essentially copy `packages/media-utils/src/utils/validate-mime-type.ts` and `packages/media-utils/src/utils/validate-file-size.ts` into this package. But if that's preferred we can of course change this.

Note that `maxUploadFileSize` is **not** currently passed to block-editor. It has a default value of `0` in `packages/block-editor/src/store/defaults.js`, but it is not referenced otherwise. This confirms it:

```
// Returns 0
wp.data.select( 'core/block-editor' ).getSettings().maxUploadFileSize

// Returns actual positive number, e.g. 314572800
wp.data.select( 'core/editor' ).getEditorSettings().maxUploadFileSize
```

So that would need to be fixed first. (In `packages/editor/src/components/provider/use-block-editor-settings.js` I think?)

Please let me know what you think.

Reply:
My preference would be to avoid introducing new settings and reuse the existing ones. These settings can also be passed down tin the "media context" if needed to be consumed by the package.

Reply:
In d3ec207c88fb8be1219ee331ebe07cdae968ee28 I now pass those settings to `block-editor` instead of the validation functions, and copied over the validation functions to `upload-media` instead of `block-editor`, as per your other suggestion in https://github.com/WordPress/gutenberg/pull/66290#discussion_r1878145369

This now means some duplicated code between `media-utils` and `upload-media`. The following functions are duplicated:

* `validateFileSize`
* `validateMimeType`
* `validateMimeTypeForUser`
	* `getMimeTypesArray`

Reply:
I've been thinking about the duplication, I think it's ok in the sense that these utils shouldn't have been defined in a package that is "WP dependent" in the first place. 

Potentially we could make `media-utils` import these from `upload-media` but it might not be possible because of the required context (settings) in `upload-media`. So maybe longer term we could just deprecated the ones in `media-utils`

Reply:
Importing from `upload-media` would cause a bit of a circular dependency.

In an ideal world there would be a shared non-WP-specific utils package both could use. `media-utils-utils` 😄

Reply:
Yeah I think the duplication is totally fine for now.